### PR TITLE
ROX-25639: Scanner Prefactor

### PR DESCRIPTION
**Pull request by Maddosaurus**
### Description

In order to move the Node Indexer service to Compliance in #12227, scanners mapper and its depending strings need to move to pkg.

### User-facing documentation

- [x] CHANGELOG update is not needed
- [x] documentation PR is not needed

### Testing and quality

- ~[ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag~
- [x] CI results are inspected

#### Automated testing

- [x] modified existing tests

#### How I validated my change

- Ensure that scanner still runs correctly local
- Ran unit tests locally
- CI

### CODE REVIEW

File: scanner/updater/manual/manual.go [hunk 1]
```diff
@@ -39,7 +40,7 @@ type Vulnerability struct {
 }
 
 const (
-	Name = "stackrox-manual"
+
```
Comment (rukletsov) [inline scanner/updater/manual/manual.go:43]: Don't think we need this blank line here

File: scanner/updater/manual/manual.go [hunk 2]
```diff
@@ -15,6 +15,7 @@ import (
 	"github.com/quay/claircore"
 	"github.com/quay/claircore/libvuln/driver"
 	"github.com/quay/zlog"
+	updater2 "github.com/stackrox/rox/pkg/scannerv4/updater"
```
Comment (rukletsov) [inline scanner/updater/manual/manual.go:18]: that's a bad name : )

File: pkg/scannerv4/mappers/mappers.go [hunk 3]
```diff
@@ -21,9 +21,9 @@ import (
 	"github.com/stackrox/rox/pkg/env"
 	"github.com/stackrox/rox/pkg/protocompat"
 	"github.com/stackrox/rox/pkg/scanners/scannerv4"
-	"github.com/stackrox/rox/scanner/enricher/fixedby"
-	"github.com/stackrox/rox/scanner/enricher/nvd"
-	"github.com/stackrox/rox/scanner/updater/manual"
+	"github.com/stackrox/rox/pkg/scannerv4/enricher/fixedby"
+	nvdp "github.com/stackrox/rox/pkg/scannerv4/enricher/nvd"
```
Comment (rukletsov) [inline pkg/scannerv4/mappers/mappers.go:25]: Can't this be _just_ `nvd`?

File: scanner/enricher/fixedby/fixedby.go [hunk 4]
```diff
@@ -28,13 +28,11 @@ import (
...
 	"github.com/quay/zlog"
+	"github.com/stackrox/rox/pkg/scannerv4/enricher/fixedby"
 )
 
 const (
-	// Type is the type of data returned from the Enricher's Enrich method.
-	Type = "message/vnd.stackrox.scannerv4.fixedby; enricher=fixedby"
-
-	// This appears above and must be the same.
+	// This appears in [fixedby.Type] and must be the same.
```
Comment (rukletsov) [inline scanner/enricher/fixedby/fixedby.go:35]:
```suggestion
	// This must be the same as the one in [fixedby.Type].
```
For an outsider, "fixedby" and "message/vnd.stackrox.scannerv4.fixedby; enricher=fixedby" are not quite the same. So who to trust here : )?

File: scanner/enricher/nvd/nvd.go [hunk 5]
```diff
@@ -33,13 +34,10 @@ var (
...
 const (
-	// Type is the type of data returned from the Enricher's Enrich method.
-	Type = `message/vnd.stackrox.scannerv4.vulnerability; enricher=nvd schema=https://csrc.nist.gov/schema/nvd/api/2.0/source_api_json_2.0.schema`
-
 	// DefaultFeeds is the default place to look for CVE feeds.
 	DefaultFeeds = `https://services.nvd.nist.gov/rest/json/cves/2.0/`
 
-	// This appears above and must be the same.
+	// This appears in [nvd.Type] and must be the same.
 	name = `nvd`
```
Comment (rukletsov) [inline scanner/enricher/nvd/nvd.go:41]: Similar to the above with `fixedby`.

File: scanner/enricher/fixedby/fixedby.go (hunk 4 above)
Comment (Maddosaurus) [reply scanner/enricher/fixedby/fixedby.go]: I've tethered the name to the type in [04591be](https://github.com/stackrox/stackrox/pull/12310/commits/04591be133538e3ee3bb71d304d15619d495de5b)

File: scanner/enricher/nvd/nvd.go (hunk 5 above)
Comment (Maddosaurus) [reply scanner/enricher/nvd/nvd.go]: I've tethered the name to the type in [04591be](https://github.com/stackrox/stackrox/pull/12310/commits/04591be133538e3ee3bb71d304d15619d495de5b), at the price of making the name public. Let's see if this raises concerns.

File: pkg/scannerv4/enricher/nvd/return.go [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package nvd
+
+const (
+	// Name of the enricher
+	Name = `nvd`
+	// Type is the type of data returned from the Enricher's Enrich method.
+	Type = `message/vnd.stackrox.scannerv4.vulnerability; enricher=` + Name + ` schema=https://csrc.nist.gov/schema/nvd/api/2.0/source_api_json_2.0.schema`
```
Comment (RTann) [inline pkg/scannerv4/enricher/nvd/return.go:7]: seems awkward, but I understand why you did it. Works for me

**Review by RTann — COMMENTED** [review]
Since it's just constants, I'm thinking we can make something like a `pkg/scannerv4/consts` package and just stuff everything in there. Then just change the names accordingly (`fixedby.Name` ==> `FixedByName`).

**Review by RTann — APPROVED** [review]
I'll approve because I don't want to block progress over something so minor. We can always adjust later (or if you want to give it a shot in this PR prior to merging 😄)

**Review by jvdm — COMMENTED** [review]
Hi, brought here to comment by @RTann, and I am just getting out of my parachute, so take my comments with some grain of salt.

1. Why are we moving Matcher-related mappers to outside `scanner/`? These mappers are considered "implementation detail" of Scanner indexing and matching, I think we would like to reserve the right to change them without having to worry about an external indexer (in this case the node). Is it for simplicity? Moving fast? The mappers are not really easy to decouple? Is the coupling between the container image indexer mapping and node indexer mapping a good thing long-term?
1. Can we use a better name for the location of these constants? `return.go` doesn't seem to communicate much. Same for the "manual updater" name. Why is it being exposed to other parts? That's minor, though.

I feel we never actually thought that parts of Scanner would be required to be used in out-of-tree services (out-of-tree = outside `scanner/`). I want to avoid spreading scanner code without a good narrative, because eventually, this will become harder  and harder to read without the context of why it was changed. Better try to put some narrative/logic behind the names, what is being moved, comments, etc.

**Comment by Maddosaurus** [comment]
@jvdm No worries, let me give you some context!

> Why are we moving Matcher-related mappers to outside scanner/?

The Node Indexer produces a `v4.IndexReport`, which (by design) should be indistinguishable from a container IndexReport. Therefore, the Node Indexer needs access to the mapping functions.
Given that both, the Indexer and the Node Indexer, should produce identical results, any change to the mapper package should impact them equally.
As to why the Node Indexer is not in `scanner/`: 
By moving the Node Indexer into the main image, we can greatly simplify the Node Scanning feature and make it more robust. We are eliminating one container and one additional image in the Collector DaemonSet, as well as a whole gRPC service.

> Can we use a better name for the location of these constants?

I will follow @RTann 's suggestion and introduce a `constants` package ([here](https://github.com/stackrox/stackrox/pull/12310#pullrequestreview-2231038751)).

**Comment by Maddosaurus** [comment]
@RTann I've moved all constants to a single package in [b5356cd](https://github.com/stackrox/stackrox/pull/12310/commits/b5356cdd7340680fa8cfb256e62927b3e4079bf6), please have a look

**Pull request merged**